File: tests/test_markdown_roundtrip.py
"""Markdown is the canonical, lossless plan format (issue #771).

The guarantee under test, as written in docs/reference/plan-format.rst:

* Opening a plan and saving it without edits leaves every byte as it was,
  except for the front-matter keys the app itself maintains (``labels``,
  ``rag``, ``version``, ``last_saved``).
* Each of those keys is changed by a single-line, in-place edit that touches
  nothing else, and re-applying the edit changes nothing.

The server's only write-back is the ``labels:`` line, returned by
``PlanService.parse`` as ``updated_plan_text`` and applied to the editor by
the page. This module runs a corpus of real plans (every bundled template
plus hand-written edge cases) through ``parse`` and checks that nothing
else moves. The browser-side writers are covered by
tests/test_markdown_roundtrip.mjs. Both run in CI
(.github/workflows/markdown-roundtrip.yml).
"""

import difflib
from pathlib import Path

import pytest

from noodle_web.plan_service import PlanService, update_front_matter_with_labels

REPO = Path(__file__).resolve().parent.parent
TEMPLATE_PLANS = sorted((REPO / "templates").glob("*/plan.md"))
FIXTURE_PLANS = sorted((REPO / "tests" / "fixtures" / "roundtrip").glob("*.md"))
CORPUS = TEMPLATE_PLANS + FIXTURE_PLANS

assert TEMPLATE_PLANS, "no bundled templates found under templates/*/plan.md"
assert FIXTURE_PLANS, "no fixtures found under tests/fixtures/roundtrip/"

APP_MAINTAINED_KEYS = ("labels", "rag", "version", "last_saved")


def _ids(paths):
    return [str(p.relative_to(REPO)) for p in paths]


def _diff(before: str, after: str) -> str:
    return "".join(
        difflib.unified_diff(
            before.splitlines(keepends=True), after.splitlines(keepends=True),
            "before", "after", n=1,
        )
    )


def _changed_lines(before: str, after: str):
    """(removed, added) line lists, so a change can be pinned to one key."""
    matcher = difflib.SequenceMatcher(a=before.split("\n"), b=after.split("\n"), autojunk=False)
    removed, added = [], []
    for op, i1, i2, j1, j2 in matcher.get_opcodes():
        if op == "equal":
            continue
        removed.extend(before.split("\n")[i1:i2])
        added.extend(after.split("\n")[j1:j2])
    return removed, added


@pytest.fixture(scope="module")
def service():
    return PlanService()


@pytest.mark.parametrize("path", CORPUS, ids=_ids(CORPUS))
def test_parse_reads_the_whole_corpus(service, path):
    """Every plan in the corpus is a valid plan; a broken fixture proves nothing."""
    result = service.parse(path.read_text(encoding="utf-8"))
    assert result.success, result.error
    assert result.tasks, f"{path.name} parsed to no tasks"


@pytest.mark.parametrize("path", CORPUS, ids=_ids(CORPUS))
def test_parse_does_not_rewrite_an_unedited_plan(service, path):
    """The server hands back the plan byte-for-byte, or changes only ``labels:``."""
    text = path.read_text(encoding="utf-8")
    result = service.parse(text)

    if result.updated_plan_text is None or result.updated_plan_text == text:
        return

    removed, added = _changed_lines(text, result.updated_plan_text)
    assert len(added) == 1 and len(removed) <= 1, (
        "parse changed more than the labels line:\n" + _diff(text, result.updated_plan_text)
    )
    assert added[0].startswith("labels:"), _diff(text, result.updated_plan_text)
    if removed:
        assert removed[0].startswith("labels:"), _diff(text, result.updated_plan_text)

    # and only because the plan used a #tag the front matter did not list
    again = service.parse(result.updated_plan_text)
    assert again.updated_plan_text in (None, result.updated_plan_text), (
        "a second parse should be a fixed point:\n"
        + _diff(result.updated_plan_text, again.updated_plan_text)
    )


@pytest.mark.parametrize("path", TEMPLATE_PLANS, ids=_ids(TEMPLATE_PLANS))
def test_bundled_templates_are_fixed_points(service, path):
    """The templates ship with a complete ``labels:`` line, so nothing at all
    changes when one is opened."""
    text = path.read_text(encoding="utf-8")
    result = service.parse(text)
    assert result.updated_plan_text in (None, text), _diff(text, result.updated_plan_text)


class TestLabelsLineIsMinimal:
    """``labels:`` is the one line the server may write; it must write it as
    narrowly as possible."""

    def test_complete_line_is_left_verbatim_whatever_its_order_or_case(self):
        plan = "---\ntitle: T\nlabels: [Zeta,  alpha ,BETA]\n---\nTask 1d #beta #zeta #Alpha\n"
        assert update_front_matter_with_labels(plan, {"beta", "zeta", "alpha"}) == plan

    def test_missing_labels_are_appended_to_the_existing_list(self):
        plan = "---\ntitle: T\nlabels: [Existing]\n---\nTask 1d #existing #newtag\n"
        out = update_front_matter_with_labels(plan, {"existing", "newtag"})
        assert out == "---\ntitle: T\nlabels: [Existing, newtag]\n---\nTask 1d #existing #newtag\n"
        assert update_front_matter_with_labels(out, {"existing", "newtag"}) == out

    def test_absent_line_is_added_as_the_last_key(self):
        plan = "---\ntitle: T\nsponsor: S\n---\nTask 1d #x\n"
        out = update_front_matter_with_labels(plan, {"x"})
        assert out == "---\ntitle: T\nsponsor: S\nlabels: [x]\n---\nTask 1d #x\n"

    def test_no_front_matter_gains_only_a_labels_block(self):
        plan = "Task 1d #x\n"
        out = update_front_matter_with_labels(plan, {"x"})
        assert out == "---\nlabels: [x]\n---\nTask 1d #x\n"
        assert update_front_matter_with_labels(out, {"x"}) == out

    def test_nothing_else_moves(self):
        plan = "---\ntitle:   spaced   \nlabels: [a]\nweird_key: 'quoted'\n---\n\n\nTask 1d #a #b   \n"
        out = update_front_matter_with_labels(plan, {"a", "b"})
        removed, added = _changed_lines(plan, out)
        assert removed == ["labels: [a]"]
        assert added == ["labels: [a, b]"]


class TestCorpusExercisesTheFormat:
    """The kitchen-sink fixture must keep covering the whole grammar, or the
    round-trip guarantee is being tested against a toy."""

    def test_kitchen_sink_covers_every_section_and_construct(self):
        text = (REPO / "tests" / "fixtures" / "roundtrip" / "kitchen-sink.md").read_text(encoding="utf-8")
        for marker in ("---highlights---", "---end-highlights---", "---raid log---", "---comms---",
                       "---budget---", "---benefits---", "---lessons learned---", "---baseline---"):
            assert marker in text, marker
        for construct in ("[depends", ":SS", "+2d", "-1d", "*+1d", "$", "^$", "/$", "{Design}", "~8h/16h",
                          "[repeats", "!!!", '"', "'", "//", "#urgent", "2026-08-03", "0d"):
            assert construct in text, construct
        for key in APP_MAINTAINED_KEYS:
            assert f"\n{key}:" in text, key

    def test_kitchen_sink_parses_the_constructs_it_claims(self, service):
        text = (REPO / "tests" / "fixtures" / "roundtrip" / "kitchen-sink.md").read_text(encoding="utf-8")
        result = service.parse(text)
        by_name = {t["name"]: t for t in result.tasks}
        assert by_name["Backend"]["dependency_types"] == {"Wireframes": "SS"}
        assert by_name["Backend"]["lag_lead"] == {"Wireframes": "+2d", "Visual design": "-1d"}
        assert by_name["Interviews"]["priority"] == "Important"
        assert by_name["Decision point"]["duration_days"] == 0
        assert by_name["Kick-off workshop"]["comment"] == "Held on site"
        assert result.front_matter.get("title", "").startswith("Kitchen Sink")
        assert result.raid_items and result.comms_items and result.highlights
        assert result.benefits_items and result.lessons_items and result.baseline_items

File: tests/test_markdown_roundtrip.mjs
/**
 * The browser side of the Markdown round-trip guarantee (issue #771).
 *
 * The page holds the plan text as-is; the only edits it makes to a plan the
 * user did not touch are the front-matter keys the app maintains:
 *
 *   rag         written after every render (persistRagToFrontMatter)
 *   version     bumped on save (incrementPlanVersion)
 *   last_saved  stamped on save (setLastSavedInFrontMatter)
 *
 * Each writer must change exactly one line, in place, leave every other byte
 * alone, and be a no-op when the value is already there. This runs those
 * writers, as the browser runs them, over the same corpus the Python guard
 * uses (tests/test_markdown_roundtrip.py).
 *
 *   node --test tests/test_markdown_roundtrip.mjs
 *
 * The writers live in static/script.js and static/version-history.js, which
 * are classic browser scripts, so their top-level function declarations are
 * lifted out by name and evaluated in a sandbox rather than required.
 */
import test from "node:test";
import assert from "node:assert/strict";
import { readFileSync, readdirSync } from "node:fs";
import { join } from "node:path";
import vm from "node:vm";

const repo = new URL("..", import.meta.url).pathname.replace(/\/$/, "");
const staticDir = join(repo, "packages", "noodle-web", "src", "noodle_web", "static");

/** Top-level `function name(` … `\n}` declarations from a classic script. */
function liftFunctions(file, names) {
  const source = readFileSync(join(staticDir, file), "utf8");
  const sandbox = {};
  for (const name of names) {
    const start = source.indexOf(`\nfunction ${name}(`);
    assert.notEqual(start, -1, `${name} not found in ${file}`);
    const end = source.indexOf("\n}\n", start);
    assert.notEqual(end, -1, `${name} in ${file} has no closing brace at column 0`);
    vm.runInNewContext(source.slice(start, end + 3), sandbox);
  }
  return sandbox;
}

const fm = {
  ...liftFunctions("script.js", [
    "getVersionFromFrontMatter", "incrementVersion", "setVersionInFrontMatter",
    "getLastSavedFromFrontMatter", "setLastSavedInFrontMatter",
  ]),
  ...liftFunctions("version-history.js", ["getRagFromFrontMatter", "setRagInFrontMatter"]),
};

const corpus = [
  ...readdirSync(join(repo, "templates"), { withFileTypes: true })
    .filter((d) => d.isDirectory())
    .map((d) => join("templates", d.name, "plan.md"))
    .filter((p) => { try { readFileSync(join(repo, p)); return true; } catch { return false; } }),
  ...readdirSync(join(repo, "tests", "fixtures", "roundtrip"))
    .filter((n) => n.endsWith(".md"))
    .map((n) => join("tests", "fixtures", "roundtrip", n)),
].sort();
assert.ok(corpus.length >= 8, "corpus should hold the templates and the fixtures");

/** Lines that differ between two texts: [removed[], added[]]. */
function changedLines(before, after) {
  const a = before.split("\n");
  const b = after.split("\n");
  // single-line in-place edits keep the line count; an inserted key adds one
  const removed = [];
  const added = [];
  if (a.length === b.length) {
    a.forEach((line, i) => { if (line !== b[i]) { removed.push(line); added.push(b[i]); } });
    return [removed, added];
  }
  let i = 0;
  while (i < a.length && a[i] === b[i]) i++;
  const tail = a.length - i;
  assert.deepEqual(a.slice(i), b.slice(b.length - tail), "change is not a single insertion");
  return [[], b.slice(i, b.length - tail)];
}

function hasFrontMatter(text) {
  return /^---\n[\s\S]*?\n---/.test(text);
}

const WRITERS = [
  {
    key: "rag",
    apply: (text) => fm.setRagInFrontMatter(text, "amber"),
    reapply: (text) => fm.setRagInFrontMatter(text, "amber"),
    expectLine: "rag: amber",
  },
  {
    key: "version",
    apply: (text) => fm.setVersionInFrontMatter(text, fm.incrementVersion(fm.getVersionFromFrontMatter(text) || "1.0")),
    reapply: (text) => fm.setVersionInFrontMatter(text, fm.getVersionFromFrontMatter(text)),
    expectLine: /^version: \d+\.\d+$/,
  },
  {
    key: "last_saved",
    apply: (text) => fm.setLastSavedInFrontMatter(text),
    reapply: (text) => fm.setLastSavedInFrontMatter(text),
    expectLine: /^last_saved: \d{4}-\d{2}-\d{2} \d{2}:\d{2}$/,
  },
];

for (const path of corpus) {
  const text = readFileSync(join(repo, path), "utf8");

  for (const w of WRITERS) {
    test(`${path}: writing ${w.key} touches only that line`, () => {
      const out = w.apply(text);

      if (!hasFrontMatter(text)) {
        // the plan had no front matter: a block holding just this key is added
        assert.ok(out.endsWith(text), "the original text must follow the new front matter unchanged");
        const head = out.slice(0, out.length - text.length).split("\n");
        assert.equal(head[0], "---");
        assert.equal(head[2], "---");
        assert.equal(head.length, 4, `front matter should hold one key: ${JSON.stringify(head)}`);
        assert.match(head[1], w.expectLine instanceof RegExp ? w.expectLine : new RegExp(`^${w.expectLine}$`));
        return;
      }

      if (out === text) {
        // the value was already there: nothing may move
        const current = text.match(new RegExp(`^${w.key}:\\s*(.*)$`, "m"));
        assert.ok(current, `${w.key} unchanged but not present in the front matter`);
        return;
      }

      const [removed, added] = changedLines(text, out);
      assert.equal(added.length, 1, `expected one changed line, got ${added.length}: ${JSON.stringify(added)}`);
      assert.match(added[0], w.expectLine instanceof RegExp ? w.expectLine : new RegExp(`^${w.expectLine}$`));
      if (removed.length) {
        assert.equal(removed.length, 1);
        assert.ok(removed[0].startsWith(`${w.key}:`), `replaced a line that was not ${w.key}: ${removed[0]}`);
      }
      // the front-matter close and everything after it are untouched
      const bodyBefore = text.slice(text.indexOf("\n---", 3));
      const bodyAfter = out.slice(out.indexOf("\n---", 3));
      assert.equal(bodyAfter, bodyBefore, "the task outline and back matter must not move");
    });

    test(`${path}: writing ${w.key} twice changes nothing`, () => {
      const once = w.apply(text);
      assert.equal(w.reapply(once), once);
    });
  }
}

test("a value that is already there is not rewritten", () => {
  const text = "---\ntitle: T\nrag: green\nversion: 2.3\n---\nTask 1d\n";
  assert.equal(fm.setRagInFrontMatter(text, "green"), text);
  assert.equal(fm.setVersionInFrontMatter(text, "2.3"), text);
  assert.equal(fm.getRagFromFrontMatter(text), "green");
  assert.equal(fm.getVersionFromFrontMatter(text), "2.3");
});

test("an odd but valid front matter is edited in place, not duplicated", () => {
  // keys with unusual spacing, a quoted value, a nested list, no trailing newline
  const text = "---\ntitle:   Spaced   \nweird: 'q'\nResources:\n- @a: A, Role\nrag: red\n---\nTask 1d";
  const out = fm.setRagInFrontMatter(text, "green");
  assert.equal(out, "---\ntitle:   Spaced   \nweird: 'q'\nResources:\n- @a: A, Role\nrag: green\n---\nTask 1d");
  assert.equal((out.match(/^---$/gm) || []).length, 2, "must not add a second front matter block");
});
